### PR TITLE
refactor(pluginserver): reset instance triggers invalidation

### PR DESCRIPTION
**Pull request by gszr**
Cherry-pick fixes https://github.com/Kong/kong-ee/pull/6417.

KAG-2859.

### CODE REVIEW

**Review by StarlightIbuki — APPROVED** [review]
This should work. But as for the EE PR, we have the same improvements that could be done:
__seq__ is not used so could be removed.

Also, a note to myself: we could  use a much simpler model, which does not require maintaining an instance list:
1. Every worker starts and maintains its instance of plugin server binary;
2. All plugin instances (for a single worker) run on the same process. It means no thread/go routine is needed

This is possible as we do not maintain states for plugin instance.

File: kong/runloop/plugin_servers/pb_rpc.lua [hunk 1]
```diff
@@ -399,19 +402,20 @@ function Rpc:handle_event(plugin_name, conf, phase)
...
-    if str_find(err_lowered, "no plugin instance")
-      or str_find(err_lowered, "closed")  then
-      kong.log.warn(err)
-      return self:handle_event(plugin_name, conf, phase)
+      kong.log.err(err_lowered)
+
+      if err_lowered == "not ready" then
+        self.reset_instance(plugin_name, conf)
+      end
+      if str_find(err_lowered, "no plugin instance")
```
Comment (tzssangglass) [inline kong/runloop/plugin_servers/pb_rpc.lua:413]: need to exit when `err_lowered` appears here?

File: kong/runloop/plugin_servers/init.lua [hunk 2]
```diff
@@ -257,22 +258,28 @@ function get_instance_id(plugin_name, conf)
   return instance_info.id
 end
 
+function reset_instances_for_plugin(plugin_name)
```
Comment (tzssangglass) [inline kong/runloop/plugin_servers/init.lua:261]: `reset_instances_by_name` is a better name?

Comment (StarlightIbuki) [reply kong/runloop/plugin_servers/init.lua:261]: note this a cherry-pick

Comment (tzssangglass) [reply kong/runloop/plugin_servers/init.lua:261]: aha

**Review by jschmid1 — APPROVED** [review]

**Pull request merged**